s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/S3ObjectSummaryIterator.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.s3.source.utils;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * Implements a ObjectSummaryIterator on an S3 bucket. Implementation reads summaries in blocks and iterates over each
+ * block. When block is empty a new block is retrieved and processed until no more data is available.
+ */
+public class S3ObjectSummaryIterator implements Iterator<S3ObjectSummary> {
+    /** The client we are using */
+    private final AmazonS3 s3Client;
+    /** The object listing from the last call to the client */
+    private ListObjectsV2Result objectListing;
+    /** The inner iterator on the object summaries. When it is empty a new one is read from object listing. */
+    private Iterator<S3ObjectSummary> innerIterator;
+
+    /** the ObjectRequest initially to start the iteration from later to retrieve more records */
+    private final ListObjectsV2Request request;
+
+    /** The last key seen by this process. This allows us to restart when a new file is dropped in the direcotry */
+    private String lastObjectSummaryKey;
+
+    /**
+     * Constructs the s3ObjectSummaryIterator based on the Amazon se client.
+     *
+     * @param s3Client
+     *            the Amazon client to read use for access.
+     * @param request
+     *            the request object that defines the starting position for the object summary retrieval.
+     */
+    @SuppressFBWarnings(value = "EI_EXPOSE_REP2", justification = "stores mutable AmazeonS3 and ListObjectsV2Request objects")
+    public S3ObjectSummaryIterator(final AmazonS3 s3Client, final ListObjectsV2Request request) {
+        this.s3Client = s3Client;
```

Comment:
This is no longer mutable in the SDK V2.X I believe so we will be able to remove this PMD suppression in the future. I am less sure about the request but also pretty sure about the request,